policysync/ipset.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2018 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package policysync
+
+import (
+	"fmt"
+
+	"github.com/projectcalico/felix/ipsets"
+	"github.com/projectcalico/felix/proto"
+	"github.com/projectcalico/libcalico-go/lib/set"
+)
+
+type ipSetInfo struct {
+	ipsets.IPSetMetadata
+	members set.Set
+}
+
+func newIPSet(update *proto.IPSetUpdate) *ipSetInfo {
+	s := &ipSetInfo{}
+
+	switch update.GetType() {
+	case proto.IPSetUpdate_IP:
+		s.Type = ipsets.IPSetTypeHashIP
+	case proto.IPSetUpdate_IP_AND_PORT:
+		s.Type = ipsets.IPSetTypeHashIPPort
+	case proto.IPSetUpdate_NET:
+		s.Type = ipsets.IPSetTypeHashNet
+	}
+
+	s.SetID = update.GetId()
+
+	// Note: We ignore MaxSize.
+
+	s.replaceMembers(update)
+	return s
+}
+
+func (s *ipSetInfo) replaceMembers(update *proto.IPSetUpdate) {
+	s.members = set.New()
+	for _, ms := range update.GetMembers() {
+		s.members.Add(s.Type.CanonicaliseMember(ms))
+	}
+}
+
+func (s *ipSetInfo) deltaUpdate(update *proto.IPSetDeltaUpdate) {
+	for _, ms := range update.GetAddedMembers() {
+		s.members.Add(s.Type.CanonicaliseMember(ms))
+	}
+	for _, ms := range update.GetRemovedMembers() {
+		s.members.Discard(s.Type.CanonicaliseMember(ms))
+	}
+}
+
+func (s *ipSetInfo) getIPSetUpdate() *proto.IPSetUpdate {
+	u := &proto.IPSetUpdate{Id: s.SetID}
+	s.members.Iter(func(item interface{}) error {
+		m := item.(fmt.Stringer)
+		u.Members = append(u.Members, m.String())
+		return nil
+	})
+	return u
+}
+
+type ruleList interface {
+	GetInboundRules() []*proto.Rule
+	GetOutboundRules() []*proto.Rule
+}
+
+func addIPSetsRuleList(rl ruleList, s map[string]bool) {
+	for _, rule := range rl.GetInboundRules() {
+		addIPSetsRule(rule, s)
+	}
+	for _, rule := range rl.GetOutboundRules() {
+		addIPSetsRule(rule, s)
+	}
+}
+
+func addIPSetsRule(r *proto.Rule, s map[string]bool) {
```

Comment:
Any way you can write a UT that will fail if a new IpSetIds field gets added without this code being updated?  Always a risk that a function like this will get forgotten in maintenance so elsewhere in Felix I've got UTs that use reflection to scan structs for new fields and compare, for example proto.Rule vs felix's ParsedRule.

For example, you could scan the proto struct for fields that end in `IpSetIds` and populate them with a slice that contains their name, then pass them through this function and check that all the names show up.  If the fields change, we'll then be forced to fix this code or the test.

Reply:
I ended up making this method public so I could test it, but don't feel great about it. What do you think?  Should the UT suite be a different package? Should I spin up a new Ginkgo suite just for this test?

Reply:
There are examples in felix where we have tests in `package name_test` for "FV" and tests in `package name` for private UTs.  You end up needing two ginkgo suite files but it does work OK.

I don't find it particularly worrying to export the odd method for test convenience, either.  It's not like we're exporting an important public API from this package.